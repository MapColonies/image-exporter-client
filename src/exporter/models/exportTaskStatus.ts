import { types, Instance } from 'mobx-state-tree';
import { Polygon } from '@turf/helpers';

export enum ExportStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'In-Progress',
  COMPLETED = 'Completed',
  FAILED = 'Failed',
}

const bbox = types.model({
  topRight: types.model({
    lat: types.number,
    lon: types.number,
  }),
  bottomLeft: types.model({
    lat: types.number,
    lon: types.number,
  }),
});

export const exportTaskStatus = types.model({
  fileName: types.string,
  sizeEst: types.number,
  realSize: types.number,
  status: types.enumeration<ExportStatus>(
    'status',
    Object.values(ExportStatus)
  ),
  maxZoom: types.number,
  polygon: types.frozen<Polygon>(),
  link: types.string,
  creationDate: types.Date,
  lastUpdateTime: types.Date,
  expirationTime: types.Date,
  progress: types.number,
  taskId: types.string,
  sourceLayer: types.string
});

export interface IExportTaskStatus extends Instance<typeof exportTaskStatus> {}

export interface IBbox extends Instance<typeof bbox> {}
