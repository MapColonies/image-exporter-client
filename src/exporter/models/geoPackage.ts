import { types, Instance } from 'mobx-state-tree';

export const GeoPackage = types.model({
  fileName: types.string,
  sizeEst: types.number,
  tilesEst: types.number,
  status: types.string,
  link: types.string,
  date: types.Date,
  progress: types.number,
});

export interface IGeoPackage extends Instance<typeof GeoPackage> {}