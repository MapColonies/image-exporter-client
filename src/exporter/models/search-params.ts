import { types, Instance } from 'mobx-state-tree';
import { Geometry } from 'geojson';

export const searchParams = types
  .model({
    geojson: types.maybe(types.frozen<Geometry>()),
  })
  .actions((self) => ({
    setLocation: function setLocation(geometry: Geometry): void {
      self.geojson = geometry;
    },

    resetLocation(): void {
      self.geojson = undefined;
    },
  }));

export interface IConflictSearchParams
  extends Instance<typeof searchParams> {}
