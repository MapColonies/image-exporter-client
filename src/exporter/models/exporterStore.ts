/* eslint-disable camelcase */
import { types, Instance, flow, getParent, getSnapshot } from 'mobx-state-tree';
import { Polygon } from '@turf/helpers';
import { AxiosRequestConfig, AxiosError } from 'axios';
import { ApiHttpResponse } from '../../common/models/api-response';
import { ResponseState } from '../../common/models/ResponseState';
import { ExportStoreError } from '../../common/models/exportStoreError';
import { getExportLayerUrl } from '../../common/helpers/layer-url';
import EXPORTER_CONFIG from '../../common/config';
// import MOCK_EXPORTED_PACKAGES from '../../__mocks-data__/exportedPackages';
import { searchParams } from './search-params';
import { IRootStore } from './rootStore';
import { IExportTaskStatus } from './exportTaskStatus';

export type ExportTaskStatusResponse = IExportTaskStatus[];
export interface ExportResult {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  data: any;
}

export interface PackageInfo {
  directoryName: string;
  packName: string;
  sizeEst: number;
  maxZoom?: number;
}

export type ExporterResponse = ApiHttpResponse<ExportResult>;

const internalError = types.model({
  request: types.maybe(types.frozen<AxiosRequestConfig>()),
  key: types.frozen<ExportStoreError>(),
});
export interface IInternalError extends Instance<typeof internalError> {}

export const exporterStore = types
  .model({
    state: types.enumeration<ResponseState>(
      'State',
      Object.values(ResponseState)
    ),
    searchParams: types.optional(searchParams, {}),
    exportedPackages: types.maybe(types.frozen<any>([])),
    errors: types.frozen<IInternalError[]>([]),
  })
  .views((self) => ({
    get root(): IRootStore {
      return getParent(self);
    },
  }))
  .actions((self) => {
    const startExportGeoPackage: (
      packInfo: PackageInfo
    ) => Promise<void> = flow(function* startExportGeoPackage(
      packInfo: PackageInfo
    ): Generator<Promise<ExporterResponse>, void, ExporterResponse> {
      self.state = ResponseState.PENDING;
      const snapshot = getSnapshot(self.searchParams);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const params: Record<string, unknown> = {};
      // Get the source layer name
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      const sourceLayer: string = EXPORTER_CONFIG.ACTIVE_LAYER_PROPERTIES.urlPatternParams.layers as string;
      // Prepare body data for request
      params.directoryName = packInfo.directoryName;
      params.fileName = packInfo.packName;
      params.sizeEst = packInfo.sizeEst;
      params.exportedLayers = [{ exportType: 'raster', url: getExportLayerUrl(), sourceLayer: sourceLayer }];
      const coordinates = (snapshot.geojson as Polygon).coordinates[0];
      params.bbox = [
        coordinates[0][0],
        coordinates[0][1],
        coordinates[2][0],
        coordinates[2][1],
      ];
      params.maxZoom = packInfo.maxZoom;

      try {
        const result = yield self.root.fetch(
          '/exportGeopackage',
          'POST',
          params
        );
        // const responseBody = result.data.data;
        self.state = ResponseState.DONE;
      } catch (e) {
        const error = e as AxiosError;
        // eslint-disable-next-line
        if (error) {
          if (
            error.response &&
            error.response.data &&
            // eslint-disable-next-line
            error.response.data.name
          ) {
            addError({
              request: error.config,
              // eslint-disable-next-line
              key: error.response.data.name as ExportStoreError,
            });
          } else {
            addError({
              request: error.config,
              key: ExportStoreError.GENERAL_ERROR,
            });
          }
        }
        self.state = ResponseState.ERROR;
      }
    });
    const getGeoPackages: () => Promise<void> = flow(
      function* getGeoPackages(): Generator<
        Promise<ExporterResponse>,
        void,
        ExportTaskStatusResponse
      > {
        try {
          self.state = ResponseState.IDLE;
          const result = yield self.root.fetch('/exportStatus', 'GET', {});
          // const result = yield Promise.resolve(MOCK_EXPORTED_PACKAGES);
          self.exportedPackages = result;
        } catch (e) {
          const error = e as AxiosError;
          self.state = ResponseState.ERROR;
          addError({
            request: error.config,
            key: ExportStoreError.GENERAL_ERROR,
          });
        }
      }
    );

    const addError = (error: IInternalError): void => {
      self.errors = [...self.errors, error];
    };

    const hasError = (key: ExportStoreError): boolean => {
      return Boolean(self.errors.find((err) => err.key === key));
    };

    const hasErrors = (): boolean => {
      const minimalLength = 0;
      return self.errors.length > minimalLength;
    };

    const cleanError = (key: ExportStoreError): boolean => {
      if (hasError(key)) {
        self.errors = self.errors.filter((err) => err.key !== key);
        return true;
      }
      return false;
    };

    const cleanErrors = (): void => {
      self.errors = [];
    };

    return {
      startExportGeoPackage,
      getGeoPackages,
      addError,
      hasError,
      hasErrors,
      cleanError,
      cleanErrors,
    };
  });

export interface IConflictsStore extends Instance<typeof exporterStore> {}
