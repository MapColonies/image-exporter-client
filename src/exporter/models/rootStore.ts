import { Method } from 'axios';
import { types, Instance, getEnv } from 'mobx-state-tree';
import { useContext, createContext } from 'react';
import { ResponseState } from '../../common/models/ResponseState';
import { exporterStore, ExporterResponse } from './exporterStore';

type FetchAction = (
  url: string,
  method: Method,
  params: Record<string, unknown>
) => Promise<ExporterResponse>;

export const baseRootStore = types
  .model({
    exporterStore: types.optional(exporterStore, {
      state: ResponseState.IDLE,
      searchParams: {},
      errors: [],
    }),
  })
  .views((self) => ({
    get fetch(): FetchAction {
      const env: { fetch: FetchAction } = getEnv(self);
      return env.fetch;
    },
  }));

export const rootStore = baseRootStore;
export interface IBaseRootStore extends Instance<typeof baseRootStore> {}
export interface IRootStore extends Instance<typeof rootStore> {}
const rootStoreContext = createContext<null | IRootStore | IBaseRootStore>(
  null
);

export const StoreProvider = rootStoreContext.Provider;
export const useStore = (): IRootStore | IBaseRootStore => {
  const store = useContext(rootStoreContext);
  if (store === null) {
    throw new Error('Store cannot be null, please add a context provider');
  }
  return store;
};
