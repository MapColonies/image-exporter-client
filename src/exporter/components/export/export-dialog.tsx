import React, { useEffect, useState } from 'react';
import { Polygon } from 'geojson';
import { useFormik } from 'formik';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  TextField,
  Button,
  Typography,
} from '@map-colonies/react-core';
import { Box } from '@map-colonies/react-components';
import { FormattedMessage, useIntl } from 'react-intl';
import { observer } from 'mobx-react-lite';
import { BBoxCorner, Corner } from '../bbox/bbox-corner-indicator';
import { getTilesCount } from '../../../common/helpers/estimated-tile-list';
import { getResolutionInMeteres } from '../../../common/helpers/zoom-resolution';
import { useDebouncedLayoutEffect } from '../../../common/hooks/debounced.hooks';
import EXPORTER_CONFIG from '../../../common/config';
import { PackageInfo } from '../../models/exporterStore';
import { ExportStoreError } from '../../../common/models/exportStoreError';
import { useStore } from '../../models/rootStore';
import { NotchLabel } from './notch-label';

const FIRST_CHAR_IDX = 0;
const DEBOUNCE_TIME = 300;
const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    spacer: {
      marginRight: '16px'
    },
    errorContainer: {
      display: 'flex',
      alignItems: 'center',
      marginRight: 'auto',
      color: theme.palette.error.main,
    },
    noBorder: {
      border: 'unset',
    },
    readOnly: {
      backgroundColor: 'transparent !important',
    },
    infoLabel: {
      width: '110px',
    },
    maxZoomField: {
      width: '40%',
    },
  })
);

// eslint-disable-next-line
const isValidPackName = (e: React.ChangeEvent<any>): boolean => {
  // eslint-disable-next-line
  const data: string = (e.nativeEvent as any).data;
  if (!data)
    return true;

  const charIdx = data.search(/[a-zA-Z0-9-_)]+/i);
  return (charIdx === FIRST_CHAR_IDX);
};

const isValidZoomValue = (zoom: number): boolean => {
  return zoom >= EXPORTER_CONFIG.EXPORT.MIN_ZOOM && zoom <= EXPORTER_CONFIG.EXPORT.MAX_ZOOM;
}

const calcPackSize = (tiles: number): number => {
  return Math.ceil(tiles * EXPORTER_CONFIG.EXPORT.AVG_TILE_SIZE_MB);
}

interface ExportDialogProps {
  isOpen: boolean;
  selectedPolygon: Polygon;
  onSetOpen: (open: boolean) => void;
  handleExport: (packInfo: PackageInfo) => void;
}

export const ExportDialog: React.FC<ExportDialogProps> = observer((props) => {
  const { exporterStore } = useStore();
  const { isOpen, onSetOpen, selectedPolygon, handleExport } = props;
  const classes = useStyle();
  const intl = useIntl();
  const formik = useFormik({
    initialValues: {
      // TODO: Need to be fixed in Shared components - this override changes the value from number to string
      bottomLeftLat: selectedPolygon.coordinates[0][0][1].toFixed(EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS),
      bottomLeftLon: selectedPolygon.coordinates[0][0][0].toFixed(EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS),
      topRightLat: selectedPolygon.coordinates[0][2][1].toFixed(EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS),
      topRightLon: selectedPolygon.coordinates[0][2][0].toFixed(EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS),
      maxZoom: EXPORTER_CONFIG.EXPORT.DEFAULT_ZOOM,
      directoryName: '',
      packageName: ''
    },
    onSubmit: values => {
      void handleExport({
        directoryName: formik.values.directoryName,
        packName: formik.values.packageName,
        maxZoom: formik.values.maxZoom,
        sizeEst: calcPackSize(numTiles),
      });
    },
  });
  // eslint-disable-next-line
  const [numTiles, setNumTiles] = useState<number>(0);
  // eslint-disable-next-line
  const [pixelSize, setPixelSize] = useState<number | string>(0);

  useDebouncedLayoutEffect(() => {
    if (isValidZoomValue(formik.values.maxZoom)) {
      const tilesCount = getTilesCount(selectedPolygon, EXPORTER_CONFIG.EXPORT.MIN_ZOOM, formik.values.maxZoom) * EXPORTER_CONFIG.EXPORT.METRIX_SET_FACTOR;
      setNumTiles(tilesCount);
      setFormErrors({ minMaxZooms: '' });
    } else {
      setFormErrors({ minMaxZooms: intl.formatMessage({ id: 'custom-bbox.form-error.zoom.invalid.text' })});
    }
  }, DEBOUNCE_TIME, [formik.values.maxZoom, selectedPolygon]);

  useDebouncedLayoutEffect(() => {
    if (isValidZoomValue(formik.values.maxZoom)) {
      const resolution = getResolutionInMeteres(formik.values.maxZoom);
      setPixelSize(resolution);
    }
  }, DEBOUNCE_TIME, [formik.values.maxZoom]);

  const [formErrors, setFormErrors] = useState({ minMaxZooms: '' });
  const [serverErrors, setServerErrors] = useState({ duplicate: '', bboxAreaForResolution: '' });

  const handleClose = (isOpened: boolean): void => {
    onSetOpen(isOpened);
  };

  // eslint-disable-next-line
  const checkName = (e: React.ChangeEvent<any>) => {

    if (serverErrors.duplicate) {
      setServerErrors({ ...serverErrors, duplicate: '' });
    }
    return isValidPackName(e) ? formik.handleChange(e) : false;
  };

  // eslint-disable-next-line
  const checkZoomLevel = (e: React.ChangeEvent<any>) => {
    if (serverErrors.bboxAreaForResolution) {
      setServerErrors({ ...serverErrors, bboxAreaForResolution: '' });
    }
    
    return formik.handleChange(e);
  };

  useEffect(() => {
    if (exporterStore.hasError(ExportStoreError.DUPLICATE_PATH)) {
      setServerErrors({ ...serverErrors, duplicate: 'export.dialog.duplicate-path.text' });
      exporterStore.cleanError(ExportStoreError.DUPLICATE_PATH);
    }
    else if(exporterStore.hasError(ExportStoreError.BBOX_TOO_SMALL_FOR_RESOLUTION)) {
      setServerErrors({ ...serverErrors, bboxAreaForResolution: 'export.dialog.bbox.resolution.validation.error.text' });
      exporterStore.cleanError(ExportStoreError.BBOX_TOO_SMALL_FOR_RESOLUTION);
    }
  }, [exporterStore, exporterStore.errors, serverErrors]);

  return (
    <Dialog open={isOpen} preventOutsideDismiss={true}>
      <DialogTitle>
        <FormattedMessage id="export.dialog.title" />
      </DialogTitle>
      <DialogContent>
        <form onSubmit={formik.handleSubmit}>
          <Box style={{ display: 'flex', marginBottom: '16px' }}>
            <TextField
              placeholder={intl.formatMessage({ id: 'export.dialog-field.directory_name.placeholder' })}
              id="directoryName"
              name="directoryName"
              type="text"
              onChange={checkName}
              value={formik.values.directoryName}
              fullwidth
            />
          </Box>
          <Box style={{ display: 'flex', marginBottom: '16px' }}>
            <TextField
              placeholder={intl.formatMessage({ id: 'export.dialog-field.package_name.placeholder' })}
              id="packageName"
              name="packageName"
              type="text"
              onChange={checkName}
              value={formik.values.packageName}
              fullwidth
            />
          </Box>
          <Box style={{ display: 'flex', marginBottom: '16px' }}>
            <TextField
              label={<NotchLabel text={intl.formatMessage({ id: 'export.dialog-field.top_right_lat.label' })} />}
              id="topRightLat"
              name="topRightLat"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.topRightLat}
              className={classes.spacer}
              readOnly
              outlined
            />
            <TextField
              label={<NotchLabel text={intl.formatMessage({ id: 'export.dialog-field.top_right_lon.label' })} />}
              id="topRightLon"
              name="topRightLon"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.topRightLon}
              className={classes.spacer}
              readOnly
              outlined
            />
            <BBoxCorner corner={Corner.TOP_RIGHT} />
          </Box>
          <Box style={{ display: 'flex', marginBottom: '16px' }}>
            <TextField
              label={<NotchLabel text={intl.formatMessage({ id: 'export.dialog-field.bottom_left_lat.label' })} />}
              id="bottomLeftLat"
              name="bottomLeftLat"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.bottomLeftLat}
              className={classes.spacer}
              readOnly
              outlined
            />
            <TextField
              label={<NotchLabel text={intl.formatMessage({ id: 'export.dialog-field.bottom_left_lon.label' })} />}
              id="bottomLeftLon"
              name="bottomLeftLon"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.bottomLeftLon}
              className={classes.spacer}
              readOnly
              outlined
            />
            <BBoxCorner corner={Corner.BOTTOM_LEFT} />
          </Box>
          <Box style={{ display: 'flex', marginBottom: '16px' }}>
            <TextField
              label={intl.formatMessage({ id: 'export.dialog-field.max_zoom.label' })}
              id="maxZoom"
              name="maxZoom"
              type="number"
              onChange={checkZoomLevel}
              value={formik.values.maxZoom}
              max={EXPORTER_CONFIG.EXPORT.MAX_ZOOM}
              min={EXPORTER_CONFIG.EXPORT.MIN_ZOOM}
              className={`${classes.spacer} ${classes.maxZoomField}`}
            />
            <Box style={{display: 'flex', alignItems: 'center'}}>
              {
                (!formErrors.minMaxZooms) ?
                  <Typography use="body2">
                    {pixelSize} {intl.formatMessage({ id: 'export.dialog-field.zoom_resolution.units.meter' })}
                  </Typography> :
                  null
              }
            </Box>
            <BBoxCorner corner={Corner.UNKNOWN} className={classes.noBorder} />
          </Box>

          <Box style={{ display: 'flex' }}>
            <Typography use="body1" className={classes.infoLabel}>
              <FormattedMessage id="export.dialog-info.label" />
            </Typography>
            <Typography use="body2">
              ~{numTiles} {intl.formatMessage({ id: 'export.dialog.tiles.text' })},&nbsp;
            </Typography>
            <Typography use="body2" style={{ marginLeft: '32px' }}>
              ~{calcPackSize(numTiles)}Mb
            </Typography>
          </Box>
          {/* <Box style={{ display: 'flex' }}>
            <Typography use="body1" className={classes.infoLabel}>
              <FormattedMessage id="export.dialog-info.link.label" /> 
            </Typography>
            <Typography use="body2" id="exportDownloadLink">
              {intl.formatMessage({ id: 'export.dialog-info.link.pattern' }, {packageName: formik.values.packageName})}
            </Typography>
          </Box> */}

          <Box style={{ display: 'flex', justifyContent: 'flex-end', marginTop: '16px', gap: '16px' }}>
            {
              (formErrors.minMaxZooms) ?
                <div className={classes.errorContainer}>
                  {`${intl.formatMessage({ id: 'general.error.label' })}: ${formErrors.minMaxZooms}`}
                </div> :
                null
            }
            {
              // Display any server error the occurred
              Object.entries(serverErrors).map(([error, value], index) => {
                return value ?
                <div key={index} className={classes.errorContainer}>
                  {`${intl.formatMessage({ id: 'general.error.label' })}: ${intl.formatMessage({ id: value })}`}
                </div> :
                null
              }
              )
            }
            <Button type="button" onClick={(): void => { handleClose(false); }}>
              <FormattedMessage id="general.cancel-btn.text" />
            </Button>
            <Button raised type="submit" disabled={!!formErrors.minMaxZooms || !formik.values.packageName || 
              !!serverErrors.duplicate || !formik.values.directoryName}>
              <FormattedMessage id="general.ok-btn.text" />
            </Button>
          </Box>
        </form>
      </DialogContent>
    </Dialog>

  );
});
