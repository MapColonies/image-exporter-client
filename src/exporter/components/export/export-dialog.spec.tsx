import React from 'react';
import { mount } from 'enzyme';
import { Polygon } from 'geojson';
import { act, waitFor } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
// eslint-disable-next-line
import '../../../__mocks__/confEnvShim';
import MESSAGES from '../../../common/i18n';
import MOCK_EXPORTED_PACKAGES from '../../../__mocks-data__/exportedPackages';
import { ExportStoreError } from '../../../common/models/exportStoreError';
import EXPORTER_CONFIG from '../../../common/config';
import { getField, updateField, updateFieldAsync } from '../../../common/test-helpers/text-field.helper.spec';
import { getButtonById } from '../../../common/test-helpers/button.helper.spec';
import { ExportTaskStatusResponse } from '../../models/exporterStore';
import { rootStore, StoreProvider } from '../../models/rootStore';
import { ExportDialog } from './export-dialog';

const setOpenFn = jest.fn();
const handleExport = jest.fn();
console.warn = jest.fn();

const exportedPackages: ExportTaskStatusResponse = MOCK_EXPORTED_PACKAGES;
const packagesFetcher = async (): Promise<ExportTaskStatusResponse> => Promise.resolve<ExportTaskStatusResponse>(exportedPackages);

const polygon: Polygon = {
  type: 'Polygon',
  coordinates: [[[32, 35], [], [31, 34], []]],
}

const fields = {
  bottomLeftLat: polygon.coordinates[0][0][1].toFixed(EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS),
  bottomLeftLon: polygon.coordinates[0][0][0].toFixed(EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS),
  topRightLat: polygon.coordinates[0][2][1].toFixed(EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS),
  topRightLon: polygon.coordinates[0][2][0].toFixed(EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS)
}

// Enzyme doesnâ€™t work properly with hooks in general, especially for `shallow` so this is the way to mock `react-intl` module.
// Enspired by https://github.com/formatjs/formatjs/issues/1477
jest.mock('react-intl', () => {
  /* eslint-disable */
  const reactIntl = require.requireActual('react-intl');
  const MESSAGES = require.requireActual('../../../common/i18n');
  const intl = reactIntl.createIntl({
    locale: 'en',
    messages: MESSAGES.default['en'],
  });

  return {
    ...reactIntl,
    useIntl: () => intl,
  };
  /* eslint-enable */
});

jest.mock('../../../common/helpers/estimated-tile-list');

describe('ExportDialog component', () => {
  it('renders correctly', async () => {

    const mockStore = rootStore.create({}, { fetch: packagesFetcher });

    const wrapper = mount(
      <StoreProvider value={mockStore}>
        <IntlProvider locale={'en'} messages={MESSAGES['en']}>
          <ExportDialog
            isOpen={true}
            onSetOpen={setOpenFn}
            selectedPolygon={polygon}
            handleExport={handleExport}
          />
        </IntlProvider>
      </StoreProvider>
    );

    await waitFor(() => {
      expect(wrapper.exists(ExportDialog)).toBeTruthy();
    });
  });

  it('Initial state of Ok button is disabled',  async () => {
    const mockStore = rootStore.create({}, { fetch: packagesFetcher });

    const wrapper = mount(
      <StoreProvider value={mockStore}>
        <IntlProvider locale={'en'} messages={MESSAGES['en']}>
          <ExportDialog
            isOpen={true}
            onSetOpen={setOpenFn}
            selectedPolygon={polygon}
            handleExport={handleExport}
          />
        </IntlProvider>
      </StoreProvider>
    );

    await waitFor(() => {
      const okButton = getButtonById(wrapper, 'general.ok-btn.text');
      expect(okButton.prop('disabled')).toBe(true);
    });
  });

  it('Passed polygon presented as bottom-left and top-right corners coordinates', async () => {
    const mockStore = rootStore.create({}, { fetch: packagesFetcher });

    const wrapper = mount(
      <StoreProvider value={mockStore}>
        <IntlProvider locale={'en'} messages={MESSAGES['en']}>
          <ExportDialog
            isOpen={true}
            onSetOpen={setOpenFn}
            selectedPolygon={polygon}
            handleExport={handleExport}
          />
        </IntlProvider>
      </StoreProvider>
    );

    for (const field in fields) {
      await waitFor(() => {
        const fieldVal = getField(wrapper, field);
        // eslint-disable-next-line
        expect(fieldVal.prop('value')).toBe((fields as any)[field]);
      });
    }

  });
  
  it('When package name and directory name are defined Ok button is enabled and download link properly generated', async () => {
    const exportPackName = 'test';
    const exportDirName = 'test';
    const mockStore = rootStore.create({}, { fetch: packagesFetcher });

    const wrapper = mount(
      <StoreProvider value={mockStore}>
        <IntlProvider locale={'en'} messages={MESSAGES['en']}>
          <ExportDialog
            isOpen={true}
            onSetOpen={setOpenFn}
            selectedPolygon={polygon}
            handleExport={handleExport}
          />
        </IntlProvider>
      </StoreProvider>
    );

    await waitFor(() => {
      updateField(wrapper, 'packageName', exportPackName);
      updateField(wrapper, 'directoryName', exportDirName);
    })

    wrapper.update();

    await waitFor(() => {
      const okButton = getButtonById(wrapper, 'general.ok-btn.text');
      expect(okButton.prop('disabled')).toBe(false);
    });
  });

  it('When all data filled and FORM submitted, handleExport triggered', async () => {
    const exportPackName = 'test';
    const exportDirName = 'test';
    const mockStore = rootStore.create({}, { fetch: packagesFetcher });

    const wrapper = mount(
      <StoreProvider value={mockStore}>
        <IntlProvider locale={'en'} messages={MESSAGES['en']}>
          <ExportDialog
            isOpen={true}
            onSetOpen={setOpenFn}
            selectedPolygon={polygon}
            handleExport={handleExport}
          />
        </IntlProvider>
      </StoreProvider>
    );

    await waitFor(() => {
      updateField(wrapper, 'packageName', exportPackName);
      updateField(wrapper, 'directoryName', exportDirName);
    })

    // eslint-disable-next-line
    await act(async () => {
      wrapper
        .find('form')
        .simulate('submit');
    });

    wrapper.update();

    await waitFor(() => {
      expect(getButtonById(wrapper, 'general.ok-btn.text').prop('disabled')).toBe(false);
      expect(handleExport).toHaveBeenCalled();
    });
  });

  it('When all data filled and FORM submitted, export fails', async () => {
    const exportPackName = 'uniqueName';
    const exportDirName = 'uniqueName';
    const polygon: Polygon = {
      type: 'Polygon',
      coordinates: [[[35.4274677973303, 32.83433188112207], [], [35.427479676963486, 32.83434257279194], []]],
    }
    const maxZoom = 3;

    const packagesFetcherFailure = async (): Promise<ExportTaskStatusResponse> => Promise.reject<ExportTaskStatusResponse>();
    const mockStore = rootStore.create({}, { fetch: packagesFetcherFailure });

    const handleExportError = jest.fn(x => {
      mockStore.exporterStore.addError({ request: {}, key: ExportStoreError.BBOX_TOO_SMALL_FOR_RESOLUTION });
    });

    const wrapper = mount(
      <StoreProvider value={mockStore}>
        <IntlProvider locale={'en'} messages={MESSAGES['en']}>
          <ExportDialog
            isOpen={true}
            onSetOpen={setOpenFn}
            selectedPolygon={polygon}
            handleExport={handleExportError}
          />
        </IntlProvider>
      </StoreProvider>
    );

    await updateFieldAsync(wrapper, 'packageName', exportPackName);
    await updateFieldAsync(wrapper, 'directoryName', exportDirName);
    await updateFieldAsync(wrapper, 'maxZoom', maxZoom);

    // eslint-disable-next-line
    await act(async () => {
      wrapper
        .find('form')
        .simulate('submit');
    });

    wrapper.update();

    await waitFor(() => {
      const errorMessage: string = MESSAGES['en']['export.dialog.bbox.resolution.validation.error.text'] as string;
      // eslint-disable-next-line
      expect(wrapper.text().includes(errorMessage)).toBe(true);
      expect(handleExportError).toHaveBeenCalled();
    });
  });

  //// TODO test to check error presentation logic.
  //// When component uses useFormik() hook internal formik state not updated when triggered onChange event.
  //// Probably formik should be used in component way rather than hook.
  //// Also preferable to test such functionality on the levevl of E2E tests.
  // it('When package name defined but one of zooms not Ok button is disabled', async () => {
  //   const exportPackName = 'test';
  //   const wrapper = shallow(
  //     <ExportDialog
  //       isOpen={true}
  //       onSetOpen={setOpenFn}
  //       selectedPolygon={polygon}
  //       handleExport={handleExport}
  //     />
  //   );

  //   await updateFieldAsync(wrapper, 'packageName', exportPackName);
  //   await updateFieldAsync(wrapper, 'minZoom', 0);

  //   wrapper.update();

  //   await waitFor(() => {
  //     const okButton = getButtonById(wrapper, 'general.ok-btn.text');
  //     expect(okButton.prop('disabled')).toBe(true);
  //   });
  // });

});
