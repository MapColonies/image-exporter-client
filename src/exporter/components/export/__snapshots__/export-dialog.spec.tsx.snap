// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportDialog component renders correctly 1`] = `
<Function
  open={true}
  preventOutsideDismiss={true}
>
  <Function>
    <FormattedMessage
      id="export.dialog.title"
    />
  </Function>
  <Function>
    <form
      onSubmit={[Function]}
    >
      <Component
        style={
          Object {
            "display": "flex",
            "marginBottom": "16px",
          }
        }
      >
        <Function
          fullwidth={true}
          id="packageName"
          name="packageName"
          onChange={[Function]}
          placeholder="Package Name"
          type="text"
          value=""
        />
      </Component>
      <Component
        style={
          Object {
            "display": "flex",
            "marginBottom": "16px",
          }
        }
      >
        <Function
          className="makeStyles-spacer-1"
          id="topRightLat"
          label={
            <NotchLabel
              text="Top Right Lat"
            />
          }
          name="topRightLat"
          onChange={[Function]}
          outlined={true}
          readOnly={true}
          type="number"
          value={34}
        />
        <Function
          className="makeStyles-spacer-1"
          id="topRightLon"
          label={
            <NotchLabel
              text="Top Right Lon"
            />
          }
          name="topRightLon"
          onChange={[Function]}
          outlined={true}
          readOnly={true}
          type="number"
          value={31}
        />
        <BBoxCorner
          corner="TOP_RIGHT"
        />
      </Component>
      <Component
        style={
          Object {
            "display": "flex",
            "marginBottom": "16px",
          }
        }
      >
        <Function
          className="makeStyles-spacer-1"
          id="bottomLeftLat"
          label={
            <NotchLabel
              text="Bottom Left Lat"
            />
          }
          name="bottomLeftLat"
          onChange={[Function]}
          outlined={true}
          readOnly={true}
          type="number"
          value={35}
        />
        <Function
          className="makeStyles-spacer-1"
          id="bottomLeftLon"
          label={
            <NotchLabel
              text="Bottom Left Lon"
            />
          }
          name="bottomLeftLon"
          onChange={[Function]}
          outlined={true}
          readOnly={true}
          type="number"
          value={32}
        />
        <BBoxCorner
          corner="BOTTOM_LEFT"
        />
      </Component>
      <Component
        style={
          Object {
            "display": "none",
            "marginBottom": "16px",
          }
        }
      >
        <Function
          className="makeStyles-spacer-1"
          id="minZoom"
          label="Min Zoom"
          name="minZoom"
          onChange={[Function]}
          type="number"
          value={1}
        />
        <Function
          className="makeStyles-spacer-1"
          id="maxZoom"
          label="Max Zoom"
          name="maxZoom"
          onChange={[Function]}
          type="number"
          value={21}
        />
        <BBoxCorner
          className="makeStyles-noBorder-3"
          corner="UNKNOWN"
        />
      </Component>
      <Component
        style={
          Object {
            "display": "flex",
          }
        }
      >
        <Function
          className="makeStyles-infoLabel-5"
          use="body1"
        >
          <FormattedMessage
            id="export.dialog-info.label"
          />
        </Function>
        <Function
          use="body2"
        >
          ~
          0
           tiles,Â 
        </Function>
        <Function
          style={
            Object {
              "marginLeft": "32px",
            }
          }
          use="body2"
        >
          ~
          0
          Mb
        </Function>
      </Component>
      <Component
        style={
          Object {
            "display": "flex",
            "gap": "16px",
            "justifyContent": "flex-end",
            "marginTop": "16px",
          }
        }
      >
        <withRipple(Function)
          onClick={[Function]}
          type="button"
        >
          <FormattedMessage
            id="general.cancel-btn.text"
          />
        </withRipple(Function)>
        <withRipple(Function)
          disabled={true}
          raised={true}
          type="submit"
        >
          <FormattedMessage
            id="general.ok-btn.text"
          />
        </withRipple(Function)>
      </Component>
    </form>
  </Function>
</Function>
`;
